device/iot-device-samples/custom-sas-token-provider-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/CustomSasTokenProviderSample.java

```diff
@@ -0,0 +1,371 @@
+    {
+        private static final String RAW_SIGNATURE_FORMAT = "%s\n%s";
+        private static final String SHARED_ACCESS_SIGNATURE_FORMAT = "SharedAccessSignature %s=%s&%s=%s&%s=%d";
+        private static final String SCOPE_FORMAT = "%s/devices/%s";
+        private static final Charset SIGNATURE_CHARSET = StandardCharsets.UTF_8;
+        private static final String EXPIRY_TIME_FIELD_KEY = "se";
+        private static final String SIGNATURE_FIELD_KEY = "sig";
+        private static final String RESOURCE_URI_FIELD_KEY = "sr";
+        private static final String HMAC_SHA_256 = "HmacSHA256";
+
+        // When deciding whether to renew SAS tokens or not, it is wise to renew proactively to avoid clock skew issues
+        // between client and server.
+        private int renewalBufferSeconds;
+        private long expiryTimeSeconds;
+        private final char[] sasToken;
+
+        /**
+         * Generate a new SAS token from your host name, device Id, and device Key.
+         * @param hostName the host name of your IoT Hub (for instance, "my-iot-hub.azure-devices.net").
+         * @param deviceId the Id of your device.
+         * @param deviceKey the primary or secondary key of your device.
+         * @param secondsToLive the number of seconds that the token will live for.
+         * @param renewalBufferSeconds the number of seconds before the token expires when this instance will recommend renewal via {{@link #shouldRenewSasToken()}}
+         */
+        public SasTokenHelper(String hostName, String deviceId, String deviceKey, int secondsToLive, int renewalBufferSeconds)
+        {
+            this.renewalBufferSeconds = renewalBufferSeconds;
+
+            try
+            {
+                // expiry time is represented by seconds since the UNIX epoch.
+                this.expiryTimeSeconds = (System.currentTimeMillis() / 1000) + secondsToLive;
+
+                String scope = buildScope(hostName, deviceId);
+
+                byte[] signature = String.format(RAW_SIGNATURE_FORMAT, scope, this.expiryTimeSeconds).getBytes(SIGNATURE_CHARSET);
+                byte[] decodedDeviceKey = Base64.decodeBase64(deviceKey);
+
+                // HMAC encrypt the signature
+                byte[] hmacEncryptedSignature = encryptSignatureHmacSha256(signature, decodedDeviceKey);
+
+                // Base64 encode the HMAC encrypted byte[]
+                byte[] base64EncodedHmacEncryptedSignature = Base64.encodeBase64(hmacEncryptedSignature);
+
+                // Convert byte[] of base64 encoded and HMAC encrypted bits to a UTF-8 String
+                String utf8Sig = new String(base64EncodedHmacEncryptedSignature, SIGNATURE_CHARSET);
+
+                // URL encode the string
+                String urlEncodedSignature = URLEncoder.encode(utf8Sig, SIGNATURE_CHARSET.name());
+
+                this.sasToken = String.format(
+                    SHARED_ACCESS_SIGNATURE_FORMAT,
+                        RESOURCE_URI_FIELD_KEY,
```

Comment:
What's with the weird tabbing going on here where some params are further in than others?

Reply:
I'll fix it